File: frontend/src/api/community.ts
import { apiClient } from './client';

export interface User {
  id: string;
  email: string;
  username: string;
}

export interface Post {
  id: string;
  user_id: string;
  user?: User;
  content: string;
  is_fact_checked: boolean;
  fact_check_result?: string;
  fact_check_status: 'pending' | 'approved' | 'flagged' | 'removed';
  likes_count: number;
  replies_count: number;
  user_liked: boolean;
  created_at: string;
  updated_at: string;
}

export interface Reply {
  id: string;
  post_id: string;
  user_id: string;
  user?: User;
  content: string;
  is_ai_response: boolean;
  created_at: string;
  updated_at: string;
}

export interface PostWithReplies {
  post: Post;
  replies: Reply[];
}

export interface CreatePostRequest {
  content: string;
}

export interface CreateReplyRequest {
  content: string;
}

export const communityApi = {
  // Get all posts (paginated)
  getPosts: async (limit = 20, offset = 0): Promise<Post[]> => {
    const response = await apiClient.get('/api/community/posts', {
      params: { limit, offset },
    });
    return response.data;
  },

  // Get a single post with replies
  getPost: async (id: string): Promise<PostWithReplies> => {
    const response = await apiClient.get(`/api/community/posts/${id}`);
    return response.data;
  },

  // Create a new post
  createPost: async (data: CreatePostRequest): Promise<Post> => {
    const response = await apiClient.post('/api/community/posts', data);
    return response.data;
  },

  // Delete a post
  deletePost: async (id: string): Promise<void> => {
    await apiClient.delete(`/api/community/posts/${id}`);
  },

  // Create a reply to a post
  createReply: async (postId: string, data: CreateReplyRequest): Promise<Reply> => {
    const response = await apiClient.post(`/api/community/posts/${postId}/replies`, data);
    return response.data;
  },

  // Toggle like on a post
  toggleLike: async (postId: string): Promise<{ liked: boolean; message: string }> => {
    const response = await apiClient.post(`/api/community/posts/${postId}/like`);
    return response.data;
  },

  // Search posts
  searchPosts: async (query: string, limit = 20, offset = 0): Promise<Post[]> => {
    const response = await apiClient.get('/api/community/posts/search', {
      params: { q: query, limit, offset },
    });
    return response.data;
  },

  // Get trending posts (top 3)
  getTrendingPosts: async (): Promise<Post[]> => {
    const response = await apiClient.get('/api/community/posts/trending');
    return response.data;
  },
};

File: backend/models/community.go
package models

import (
	"time"

	"github.com/google/uuid"
	"gorm.io/gorm"
)

// Post represents a user-generated post in the community
type Post struct {
	ID                uuid.UUID `gorm:"type:uuid;primary_key;default:uuid_generate_v4()" json:"id"`
	UserID            uuid.UUID `gorm:"type:uuid;not null" json:"user_id"`
	User              *User     `gorm:"foreignKey:UserID" json:"user,omitempty"`
	Content           string    `gorm:"type:text;not null" json:"content"`
	IsFactChecked     bool      `gorm:"default:false" json:"is_fact_checked"`
	FactCheckResult   string    `gorm:"type:text" json:"fact_check_result,omitempty"`
	FactCheckStatus   string    `gorm:"type:varchar(50);default:'pending'" json:"fact_check_status"` // pending, approved, flagged, removed
	LikesCount        int       `gorm:"-" json:"likes_count"`                                         // Computed field
	RepliesCount      int       `gorm:"-" json:"replies_count"`                                       // Computed field
	UserLiked         bool      `gorm:"-" json:"user_liked"`                                          // Computed field for current user
	CreatedAt         time.Time `json:"created_at"`
	UpdatedAt         time.Time `json:"updated_at"`
}

// Reply represents a response to a post
type Reply struct {
	ID           uuid.UUID `gorm:"type:uuid;primary_key;default:uuid_generate_v4()" json:"id"`
	PostID       uuid.UUID `gorm:"type:uuid;not null" json:"post_id"`
	UserID       uuid.UUID `gorm:"type:uuid;not null" json:"user_id"`
	User         *User     `gorm:"foreignKey:UserID" json:"user,omitempty"`
	Content      string    `gorm:"type:text;not null" json:"content"`
	IsAIResponse bool      `gorm:"default:false" json:"is_ai_response"`
	CreatedAt    time.Time `json:"created_at"`
	UpdatedAt    time.Time `json:"updated_at"`
}

// PostLike represents a like/upvote on a post
type PostLike struct {
	ID        uuid.UUID `gorm:"type:uuid;primary_key;default:uuid_generate_v4()" json:"id"`
	PostID    uuid.UUID `gorm:"type:uuid;not null;uniqueIndex:idx_post_user_like" json:"post_id"`
	UserID    uuid.UUID `gorm:"type:uuid;not null;uniqueIndex:idx_post_user_like" json:"user_id"`
	CreatedAt time.Time `json:"created_at"`
}

// PostWithDetails includes post with related data
type PostWithDetails struct {
	Post
	Replies []Reply `json:"replies,omitempty"`
}

// TrendingPost represents a post in the trending list
type TrendingPost struct {
	Post
	Rank int `json:"rank"`
}

// BeforeCreate hooks for UUID generation
func (p *Post) BeforeCreate(tx *gorm.DB) error {
	if p.ID == uuid.Nil {
		p.ID = uuid.New()
	}
	if p.FactCheckStatus == "" {
		p.FactCheckStatus = "pending"
	}
	return nil
}

func (r *Reply) BeforeCreate(tx *gorm.DB) error {
	if r.ID == uuid.Nil {
		r.ID = uuid.New()
	}
	return nil
}

func (pl *PostLike) BeforeCreate(tx *gorm.DB) error {
	if pl.ID == uuid.Nil {
		pl.ID = uuid.New()
	}
	return nil
}

// Request/Response structs
type CreatePostRequest struct {
	Content string `json:"content" binding:"required,min=1,max=5000"`
}

type CreateReplyRequest struct {
	Content string `json:"content" binding:"required,min=1,max=2000"`
}

type SearchPostsRequest struct {
	Query  string `json:"query" form:"query" binding:"required,min=1"`
	Limit  int    `json:"limit" form:"limit"`
	Offset int    `json:"offset" form:"offset"`
}
